core/trino-main/src/main/java/io/trino/sql/gen/PageFunctionCompiler.java

```diff
@@ -203,11 +204,12 @@ private Supplier<PageProjection> compileProjectionInternal(RowExpression project
             throw new TrinoException(COMPILER_ERROR, e);
         }
 
+        MethodHandle pageProjectionConstructor = constructorMethodHandle(pageProjectionWorkClass, BlockBuilder.class, ConnectorSession.class, Page.class, SelectedPositions.class);
         return () -> new GeneratedPageProjection(
                 result.getRewrittenExpression(),
                 isDeterministic(result.getRewrittenExpression()),
                 result.getInputChannels(),
-                constructorMethodHandle(pageProjectionWorkClass, BlockBuilder.class, ConnectorSession.class, Page.class, SelectedPositions.class));
+                pageProjectionConstructor);
```

Comment:
PageProcessors are not thread-safe (some function implementations can have state), so each instance of an operator needs to get their own instance of a page processor.

Reply:
> maybe in a different ClassLoader context?
I don't think so. Anyway, in that particular case it does not matter on my eye: `LambdaForm`s are generated in fly by JVM internal class loader / machinery and I don't think so that LF can be generated by user-provided class-loader :)

Reply:
> PageProcessors are not thread-safe (some function implementations can have state)

Makes sense. I see that actually that `GeneratedPageProjection` has state

>> maybe in a different ClassLoader context?
>
>I don't think so. Anyway, in that particular case it does not matter on my eye: LambdaForms are generated in fly by JVM internal class loader / machinery and I don't think so that LF can be generated by user-provided class-loader.

I theoretically see two MH targeting the same class loaded by two different class loaders e.g. to have a different jit profile or some other benefit of class loader isolation. Like you and Martin said it's not the case here so looks good.